src/main/java/htsjdk/samtools/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,674 @@
+
+    /**
+     * Bases per line to be applied to the sequence that is been currently appended to the output.
+     */
+    private int currentBasesPerLine;
+
+    /**
+     * Holds the number of bases in the current output line.
+     */
+    private int currentLineBasesCount;
+
+    /**
+     * Holds the number of bases so far appended for the current sequence.
+     */
+    private long currentBasesCount;
+
+    /**
+     * Holds the FASTA output file offset for the current sequence.
+     */
+    private long currentSequenceOffset;
+
+    /**
+     * Holds the name of the sequence that is been appended currently.
+     */
+    private String currentSequenceName;
+
+    /**
+     * Flag indicating whether this writer has been already closed.
+     */
+    private boolean closed;
+
+    /**
+     * Creates a reference FASTA file writer (private...use the builder: {@link FastaReferenceWriterBuilder}.
+     * <p>
+     * You can specify a specific output stream to each file: the main fasta output, its index and its dictionary.
+     * </p>
+     *
+     * @param fastaOutput the output fasta file path.
+     * @param indexOutput the output stream to the index file, if requested, {@code null} if none should be generated.
+     * @param dictOutput  the output stream to the dictFile, if requested, {@code null} if none should be generated.
+     * @throws IllegalArgumentException if {@code fastaFile} is {@code null} or {@code basesPerLine} is 0 or negative.
+     */
+    protected FastaReferenceWriter(final int basesPerLine,
+                                final OutputStream fastaOutput,
+                                final OutputStream indexOutput,
+                                final OutputStream dictOutput) {
+        this.defaultBasePerLine = basesPerLine;
+        this.fastaStream = new CountingOutputStream(fastaOutput);
+        this.indexWriter = indexOutput == null ? NullWriter.NULL_WRITER : new OutputStreamWriter(indexOutput, CHARSET);
+        final BufferedWriter dictWriter = new BufferedWriter(dictOutput == null ? NullWriter.NULL_WRITER : new OutputStreamWriter(dictOutput, CHARSET));
+        this.dictWriter = dictWriter;
+        this.dictCodec = new SAMSequenceDictionaryCodec(dictWriter);
+        this.dictCodec.encodeHeaderLine(false);
+    }
+
+    // checks that a sequence name is valid.
+    private static void checkSequenceName(final String name) {
+        ValidationUtils.nonEmpty(name, "Sequence name");
+
+        for (int i = 0; i < name.length(); i++) {
+            final char ch = name.charAt(i);
+            if (Character.isWhitespace(ch)) {
+                throw new IllegalArgumentException("the input name contains blank characters: '" + name + "'");
+            } else if (Character.isISOControl(ch)) {
+                throw new IllegalArgumentException("the input name contains control characters: '" + name + "'");
+            }
+        }
```

Comment:
If you stored `SAMSequenceRecord`s this would take care of itself!

Reply:
yeah, but I can't use that and still get things like description (which SAMSequenceRecord doesn't support) and also incremental buildup of a sequence..